plugins/usbdmx/DMXCreator512Basic.cpp

```diff
@@ -0,0 +1,287 @@
+/*
+ * This program is free software; you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License as published by
+ * the Free Software Foundation; either version 2 of the License, or
+ * (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU Library General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program; if not, write to the Free Software
+ * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA.
+ *
+ * DMXCreator512Basic.cpp
+ * The synchronous and asynchronous DMXCreator512Basic widgets.
+ * Copyright (C) 2016 Florian Edelmann
+ */
+
+#include "plugins/usbdmx/DMXCreator512Basic.h"
+
+#include <unistd.h>
+#include <string.h>
+#include <string>
+
+#include "libs/usb/LibUsbAdaptor.h"
+#include "ola/Logging.h"
+#include "ola/Constants.h"
+#include "plugins/usbdmx/AsyncUsbSender.h"
+#include "plugins/usbdmx/ThreadedUsbSender.h"
+
+/**
+ * DMXCreator 512 Basic expects two or three URB packets for each change:
+ * 1. A constant byte string to endpoint 1 that indicates if we want to transmit
+ *    the the full universe or only the first half.
+ * 2. The actual DMX data (channels 1..256) to endpoint 2.
+ * 3. The actual DMX data (channels 257..512) to endpoint 2. (optional)
+ */
+
+namespace ola {
+namespace plugin {
+namespace usbdmx {
+
+using std::string;
+using ola::usb::LibUsbAdaptor;
+
+namespace {
+
+static const unsigned char ENDPOINT_1 = 0x01;
+static const unsigned char ENDPOINT_2 = 0x02;
+static const unsigned int URB_TIMEOUT_MS = 50;
+static const unsigned int CHANNELS_PER_PACKET = 256;
+
+// if we only wanted to send the first half of the universe, the last byte would
+// be 0x01
+static const unsigned char status_buffer[6] = {
+  0x80, 0x01, 0x00, 0x00, 0x00, 0x02
+};
+
+typedef enum {
+  STATE_SEND_STATUS,
+  STATE_SEND_FIRST_HALF,
+  STATE_SEND_SECOND_HALF
+} SendState;
+
+}  // namespace
+
+// DMXCreator512BasicThreadedSender
+// -----------------------------------------------------------------------------
+
+/**
+ * Sends messages to a DMXCreator 512 Basic device in a separate thread.
+ */
+class DMXCreator512BasicThreadedSender: public ThreadedUsbSender {
+ public:
+  DMXCreator512BasicThreadedSender(LibUsbAdaptor *adaptor,
+                                   libusb_device *usb_device,
+                                   libusb_device_handle *handle)
+      : ThreadedUsbSender(usb_device, handle),
+        m_adaptor(adaptor) {
+    m_dmx_buffer = new DmxBuffer();
+    m_dmx_buffer_1 = new uint8_t[CHANNELS_PER_PACKET];
+    m_dmx_buffer_2 = new uint8_t[CHANNELS_PER_PACKET];
+  }
+
+ private:
+  DmxBuffer *m_dmx_buffer;
+  uint8_t *m_dmx_buffer_1;
```

Comment:
these should be uint8_t m_dmx_buffer[CHANNELS_PER_PACKET] to avoid having to manually manage memory - as it stands we're leaking the arrays & the dmx buffer.